### PR TITLE
Fix complexity analysis

### LINKED ISSUE #2373: Assertion violation in ConcurrentFlyweight.h `Idx < SlotCount.load(std::memory_order_relaxed)'
Hi!

I have a huge souffle-file that consistently triggers the following assertion violation in `ConcurrentFlyweight`:

```
/nix/store/99qgqy9ar3aa4969qg8x10pb6rvxi619-souffle-2.3/include/souffle/datastructure/ConcurrentFlyweight.h:290: const Key& souffle::ConcurrentFlyweight<LanesPolicy, Key, Hash, KeyEqual, KeyFactory>::fetch(souffle::ConcurrentFlyweight<LanesPolicy, Key, Hash, KeyEqual, KeyFactory>::lane_id, souffle::ConcurrentFlyweight<LanesPolicy, Key, Hash, KeyEqual, KeyFactory>::index_type) const [with LanesPolicy = souffle::ConcurrentLanes; Key = std::__cxx11::basic_string<char>; Hash = std::hash<std::__cxx11::basic_string<char> >; KeyEqual = std::equal_to<std::__cxx11::basic_string<char> >; KeyFactory = souffle::details::Factory<std::__cxx11::basic_string<char> >; souffle::ConcurrentFlyweight<LanesPolicy, Key, Hash, KeyEqual, KeyFactory>::lane_id = long unsigned int; souffle::ConcurrentFlyweight<LanesPolicy, Key, Hash, KeyEqual, KeyFactory>::index_type = long unsigned int]: Assertion `Idx < SlotCount.load(std::memory_order_relaxed)' failed.
```

I'm not quite sure how to trigger the bug in a smaller example that would be useful for you to debug, but this is what I've found out so far:

* the souffle rule that causes the assertion violation is

```souffle
_MatchExpBranch24(0, $V(?_p), ?_x, $Mem(?_s0, ?_s1, ?_s2), ?_addr, ?_mem, ?_v, $Mem($T(), $T(), $T())) :-
  (?_p != @intToI256("20")),
  (?_p != @intToI256("40")),
  _MatchExpBranch24((0+1), $V(?_p), ?_x, $Mem(?_s0, ?_s1, ?_s2), ?_addr, ?_mem, ?_v, _),
  (1-@i256lt(@intToI256("60"),?_p)) != 0,
  (?_p != @intToI256("0")).
```

* the assertion violation is called from when calling `symbolTable->decode` in a stateful external functor (`i256lt`), as can be seen in the stack trace below:

```
#0  0x00007ffff78f5bc7 in __pthread_kill_implementation () from /nix/store/hsk71z8admvgykn7vzjy11dfnar9f4r1-glibc-2.35-163/lib/libc.so.6
#1  0x00007ffff78a8b46 in raise () from /nix/store/hsk71z8admvgykn7vzjy11dfnar9f4r1-glibc-2.35-163/lib/libc.so.6
#2  0x00007ffff78934b5 in abort () from /nix/store/hsk71z8admvgykn7vzjy11dfnar9f4r1-glibc-2.35-163/lib/libc.so.6
#3  0x00007ffff78933d9 in __assert_fail_base.cold.0 () from /nix/store/hsk71z8admvgykn7vzjy11dfnar9f4r1-glibc-2.35-163/lib/libc.so.6
#4  0x00007ffff78a17b6 in __assert_fail () from /nix/store/hsk71z8admvgykn7vzjy11dfnar9f4r1-glibc-2.35-163/lib/libc.so.6
#5  0x0000000000439559 in souffle::SymbolTableImpl::decode[abi:cxx11](int) const ()
#6  0x00007ffff7dd2a59 in horst::compare_symbols_gmp(souffle::SymbolTable*, int, int) () from ./libfunctors.so
#7  0x00007ffff7dd3341 in horst::compare_symbols_gmp_cached(souffle::SymbolTable*, int, int, std::unordered_map<unsigned long, int, std::hash<unsigned long>, std::equal_to<unsigned long>, std::allocator<std::pair<unsigned long const, int> > >&) () from ./libfunctors.so
#8  0x00007ffff7dd14e4 in i256lt () from ./libfunctors.so
#9  0x000000000088f3c1 in souffle::Sf_rules_orig::subroutine_10(std::vector<int, std::allocator<int> > const&, std::vector<int, std::allocator<int> >&)::{lambda()#1585}::operator()() const ()
#10 0x000000000094f311 in souffle::Sf_rules_orig::subroutine_10(std::vector<int, std::allocator<int> > const&, std::vector<int, std::allocator<int> >&) ()
#11 0x000000000099fb5b in souffle::Sf_rules_orig::runFunction(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool, bool) ()
#12 0x00000000009a0f3e in souffle::Sf_rules_orig::runAll(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool, bool) ()
```
* the assertion violation is triggered with deterministic timing (after ~14 seconds on my machine)
* the assertion violation is dependent on the order of premises in the generated c++ file. When manually changing the order, it does not get triggered anymore (I don't know how relevant this is, maybe the call to  `i256lt` that triggers the assertion violation is just not executed because of short-circuit-evaluation.) 

```c++

//original
if( (ramBitCast<RamDomain>(env1[1]) != ramBitCast<RamDomain>(functors::intToI256(&symTable, &recordTable,RamSigned(22)))) && (ramBitCast<RamDomain>(env1[0]) == ramBitCast<RamDomain>(RamSigned(23))) && (ramBitCast<RamDomain>((ramBitCast<RamSigned>(RamSigned(1)) - ramBitCast<RamSigned>(functors::i256lt(&symTable, &recordTable,functors::intToI256(&symTable, &recordTable,RamSigned(21)),env1[1])))) != ramBitCast<RamDomain>(RamSigned(0))) && (ramBitCast<RamDomain>(env1[1]) != ramBitCast<RamDomain>(functors::intToI256(&symTable, &recordTable,RamSigned(23)))) && (ramBitCast<RamDomain>(env1[1]) != ramBitCast<RamDomain>(functors::intToI256(&symTable, &recordTable,RamSigned(0)))))

//manually changed 
if( (ramBitCast<RamDomain>(env1[0]) == ramBitCast<RamDomain>(RamSigned(23))) && (ramBitCast<RamDomain>((ramBitCast<RamSigned>(RamSigned(1)) - ramBitCast<RamSigned>(functors::i256lt(&symTable, &recordTable,functors::intToI256(&symTable, &recordTable,RamSigned(21)),env1[1])))) != ramBitCast<RamDomain>(RamSigned(0))) && (ramBitCast<RamDomain>(env1[1]) != ramBitCast<RamDomain>(functors::intToI256(&symTable, &recordTable,RamSigned(22)))) && (ramBitCast<RamDomain>(env1[1]) != ramBitCast<RamDomain>(functors::intToI256(&symTable, &recordTable,RamSigned(23)))) && (ramBitCast<RamDomain>(env1[1]) != ramBitCast<RamDomain>(functors::intToI256(&symTable, &recordTable,RamSigned(0)))))
```

* `i256lt` is called multiple time without triggering the violation (before eventually triggering the violation).

I'm not quite sure how I could debug this and would be happy to get some pointers. I could also share the the souffle-file (preferably privately) but I'm not sure if the bug reproduces (also setting everything up might be a quite cumbersome as the external functor shared object need to link a library etc.).

So far I have reproduced to bug in releases `2.2` and `2.3`

Thanks in advance!

### PR DESCRIPTION
**Pull request by quentin**
The complexity analysis was ignoring intrinsics operations that can contain functor calls. It had an impact on the reordering of conditions that would move a functor call too early when it is within an intrinsic operation.

fix #2373

### CODE REVIEW

**Comment by XiaowenHu96** [comment]
If I understand the problem correctly, the optimizer was not considering the dependence between branch check and functor call. And the fix is to give functor call a larger complexity. If so, can we give it an INT_MAX to make it clear that we want functor calls to be put at last?

**Comment by quentin** [comment]
I agree, the current fix is not sufficient in some situations and associating the maximum complexity rank to functor call should solve the issue entirely.

File: src/ram/analysis/Complexity.cpp [hunk 1]
```diff
@@ -62,7 +62,7 @@ int ComplexityAnalysis::getComplexity(const Node* node) const {
         }
 
         int visit_(type_identity<UserDefinedOperator>, const UserDefinedOperator&) override {
-            return 10;
```
Comment (XiaowenHu96) [inline src/ram/analysis/Complexity.cpp:65]: As per the comment, make it the maximum value of Int.

**Review by XiaowenHu96 — CHANGES_REQUESTED** [review]

File: src/ram/analysis/Complexity.cpp (hunk 1 above)
Comment (quentin) [reply src/ram/analysis/Complexity.cpp:65]: Done.

**Review by XiaowenHu96 — APPROVED** [review]
Thanks!

**Pull request merged**